### PR TITLE
2D Rope into develop

### LINKED ISSUE #1109: Update positional encodings in model to be purely local
### Describe the task. It can be a feature, documentation, etc.

Currently, various positional encodings assuming a fixed healpix grid and use absolute coordinates. This should be changed to relative positional encodings to avoid grid artifacts--if possible and the skill is not degraded by this.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by kctezcan**
## Description

This is @csjfwang 's implementation of 2D rope, coming from this orignal PR: https://github.com/ecmwf/WeatherGenerator/pull/1445

## Issue Number

Closes https://github.com/ecmwf/WeatherGenerator/pull/1445
Closes #1109 

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 1]
```diff
@@ -808,4 +808,4 @@ def worker_workset(self):
                 + f" : dataset [{local_start},{local_end}) : [{iter_start},{iter_end})"
             )
 
-        return iter_start, iter_end
\ No newline at end of file
+        return iter_start, iter_end
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py:811]: Can we revert this

File: src/weathergen/model/engines.py [hunk 2]
```diff
@@ -301,9 +301,12 @@ def __init__(self, cf: Config, num_healpix_cells: int) -> None:
                 )
             )
 
-    def forward(self, tokens, use_reentrant):
+    def forward(self, tokens, coords=None, use_reentrant=False):
         for block in self.ae_aggregation_blocks:
-            tokens = checkpoint(block, tokens, use_reentrant=use_reentrant)
+            if isinstance(block, MultiSelfAttentionHead | MultiSelfAttentionHeadLocal):
```
Comment (clessig) [inline src/weathergen/model/engines.py:306]: We could avoid the branches if we would follow the approach in the MLP:

https://github.com/ecmwf/WeatherGenerator/blob/29e2fe2c6e213e262480edd3fabfa3d4200e42a3/src/weathergen/model/layers.py#L82

File: src/weathergen/model/engines.py [hunk 3]
```diff
@@ -373,9 +376,12 @@ def __init__(self, cf: Config, num_healpix_cells: int) -> None:
                 torch.nn.LayerNorm(self.cf.ae_global_dim_embed, elementwise_affine=False)
             )
 
-    def forward(self, tokens, use_reentrant):
+    def forward(self, tokens, coords=None, use_reentrant=False):
         for block in self.ae_global_blocks:
-            tokens = checkpoint(block, tokens, use_reentrant=use_reentrant)
+            if isinstance(block, MultiSelfAttentionHead | MultiSelfAttentionHeadLocal):
```
Comment (clessig) [inline src/weathergen/model/engines.py:381]: Same as above: if-statement is not necessary

File: src/weathergen/model/engines.py (hunk 2 above)
Comment (clessig) [reply src/weathergen/model/engines.py:306]: Or some compromise with *args and **kwargs

File: src/weathergen/model/engines.py [hunk 4]
```diff
@@ -463,13 +469,16 @@ def init_weights_final(m):
         for block in self.fe_blocks:
             block.apply(init_weights_final)
 
-    def forward(self, tokens, fstep):
+    def forward(self, tokens, fstep, coords=None):
         aux_info = None
         for _b_idx, block in enumerate(self.fe_blocks):
             if isinstance(block, torch.nn.modules.normalization.LayerNorm):
                 tokens = block(tokens)
+            elif isinstance(block, MultiSelfAttentionHead | MultiSelfAttentionHeadLocal):
```
Comment (clessig) [inline src/weathergen/model/engines.py:477]: Same as above

File: src/weathergen/model/model.py [hunk 5]
```diff
@@ -101,6 +102,21 @@ def __init__(self, cf) -> None:
         )
         self.pe_global = torch.nn.Parameter(pe, requires_grad=False)
 
+        ### ROPE COORDS ###
```
Comment (clessig) [inline src/weathergen/model/model.py:105]: standard comment: 

```
# rope coords
```

File: src/weathergen/model/attention.py [hunk 6]
```diff
@@ -539,6 +543,9 @@ def forward(self, x, ada_ln_aux=None):
         ks = self.lnorm_k(self.proj_heads_k(x).reshape(s)).to(self.dtype)
         vs = self.proj_heads_v(x).reshape(s).to(self.dtype)
 
+        if coords is not None:
```
Comment (clessig) [inline src/weathergen/model/attention.py:546]: MultiSelfAttenionHead is a vanilla self attention block. The 2D rope positional encoding makes strong assumption of the structure of the qs and ks. At the very least, we need some documentation on what the option means and when it makes sense. Not sure if there's another good way to ensure proper use of this option (one could have a separate class but then one should make sure there's no code duplication).

**Review by clessig — COMMENTED** [review]
We can simplify the code if we change the syntax of the forward fct of the attention blocks slightly. This should probably be done consistently for all heads, even those without 2D rope.

File: src/weathergen/model/engines.py (hunk 2 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/engines.py:306]: Personal preference is the if statement

File: src/weathergen/model/engines.py (hunk 3 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/engines.py:381]: but cleaner to read imho

File: src/weathergen/model/model.py [hunk 7]
```diff
@@ -165,28 +182,55 @@ def reset_parameters(self, cf: Config) -> "ModelParams":
...
-        self.pe_global.data[..., 0::2] = 0.5 * torch.sin(
-            torch.outer(8 * torch.arange(cf.ae_local_num_queries, device=self.pe_global.device), xs)
-        )
-        self.pe_global.data[..., 0::2] += (
-            torch.sin(
-                torch.outer(torch.arange(self.num_healpix_cells, device=self.pe_global.device), xs)
+
+        if self.rope_2D:
+            # Precompute per-cell center coordinates (lat, lon in radians) for 2D RoPE.
+            # Shape: (num_healpix_cells, ae_local_num_queries, 2)
```
Comment (sophie-xhonneux) [inline src/weathergen/model/model.py:188]: Why isn't the shape `(num_healpix_cells,  2)`?

File: src/weathergen/model/model.py [hunk 8]
```diff
@@ -101,6 +102,21 @@ def __init__(self, cf) -> None:
         )
         self.pe_global = torch.nn.Parameter(pe, requires_grad=False)
 
+        ### ROPE COORDS ###
+        self.rope_2D = cf.get("rope_2D", False)
+        if self.rope_2D:
+            self.rope_coords = torch.nn.Parameter(
```
Comment (sophie-xhonneux) [inline src/weathergen/model/model.py:108]: I think this should be a registered buffer, not parameters

Comment (sophie-xhonneux) [reply src/weathergen/model/model.py:108]: https://docs.pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer

File: src/weathergen/model/positional_encoding.py [hunk 9]
```diff
@@ -94,3 +94,83 @@ def positional_encoding_harmonic_coord(x, lats, lons):
...
+    cos = cos.unsqueeze(unsqueeze_dim).to(dtype=q.dtype)
+    sin = sin.unsqueeze(unsqueeze_dim).to(dtype=q.dtype)
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+
+    return q_embed, k_embed
+
+
+####################################################################################################
+def rotary_embedding_2d(coords, dim, base=10000.0):
```
Comment (sophie-xhonneux) [inline src/weathergen/model/positional_encoding.py:135]: There should be an option to learn the frequencies, this proposed in the original RoPE (https://arxiv.org/pdf/2403.13298v1 section 3.2 at the end) or we should have a reference explaining why not

File: src/weathergen/model/positional_encoding.py [hunk 10]
```diff
@@ -94,3 +94,83 @@ def positional_encoding_harmonic_coord(x, lats, lons):
...
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+    """Applies Rotary Position Embedding to the query and key tensors.
+
+    Args:
+        q: Query tensor.
+        k: Key tensor.
+        cos: Cosine embedding tensor.
+        sin: Sine embedding tensor.
+        position_ids: Deprecated and unused; present for API compatibility.
```
Comment (sophie-xhonneux) [inline src/weathergen/model/positional_encoding.py:122]: Let's delete, we have no requirement to be backwards compatible

File: src/weathergen/model/positional_encoding.py [hunk 11]
```diff
@@ -94,3 +94,83 @@ def positional_encoding_harmonic_coord(x, lats, lons):
...
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
```
Comment (sophie-xhonneux) [inline src/weathergen/model/positional_encoding.py:114]: This looks like 1D-rope to me?

Comment (sophie-xhonneux) [reply src/weathergen/model/positional_encoding.py:114]: Can we please label the function as such?

File: src/weathergen/model/positional_encoding.py [hunk 12]
```diff
@@ -94,3 +94,83 @@ def positional_encoding_harmonic_coord(x, lats, lons):
...
+        coords: Tensor of shape (..., 2) with coordinates in radians (lat, lon).
+        dim: Head dimension to encode; must be divisible by 4.
+        base: RoPE base frequency.
+
+    Returns:
+        Tuple of (cos, sin) tensors with shape (..., dim).
+    """
+
+    if coords.shape[-1] != 2:
+        raise ValueError(f"coords last dimension must be 2 (lat, lon); got {coords.shape[-1]}")
```
Comment (sophie-xhonneux) [inline src/weathergen/model/positional_encoding.py:148]: can we just make these asserts? it is cleaner to read

File: src/weathergen/model/attention.py [hunk 13]
```diff
@@ -242,7 +243,7 @@ def mask_block_local(batch, head, idx_q, idx_kv):
         # compile for efficiency
         self.flex_attention = torch.compile(flex_attention, dynamic=False)
 
-    def forward(self, x, ada_ln_aux=None):
+    def forward(self, x, coords=None, ada_ln_aux=None):
```
Comment (sophie-xhonneux) [inline src/weathergen/model/attention.py:246]: From a code perspective I prefer if we decide at init time if this attention module has 2D rope, because a) we may wish to learn the frequencies theta_t and b) we should not use an attention module trained with 2D-rope without it (or at least be very intentional about it)

**Review by sophie-xhonneux — COMMENTED** [review]
Did a first review of the code and left some comments that should be addressed.

Will check the correctness next
